src/poly1305.rs

```diff
@@ -16,31 +16,115 @@
-        unsafe { GFp_poly1305_init(&mut ctx.state, &key.bytes); }
+        check_block_state_layout();
+
+        let set_fns = init(AlignedBlockState::new(&mut ctx.opaque).aligned_buf(), &key.bytes, &mut ctx.func) == 0;
+        // TODO XXX: It seems at least some implementations |poly1305_init| always
+        // return the same value, so this conditional logic isn't always necessary.
+        // And, for platforms that have such conditional logic also in the ASM code,
+        // it seems it would be better to move the conditional logic out of the asm
+        // and into the higher-level code.
+        if !set_fns {
+            ctx.func.blocks_fn = GFp_poly1305_blocks;
+            ctx.func.emit_fn = GFp_poly1305_emit;
+        }
+
+        ctx.nonce = [
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[16..20], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[20..24], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[24..28], 4).unwrap()),
+            u32_from_le_u8(slice_as_array_ref!(&key.bytes[28..32], 4).unwrap()),
+        ];
         ctx
     }
 
-    #[inline]
-    pub fn update(&mut self, input: &[u8]) {
-        unsafe {
-            GFp_poly1305_update(&mut self.state, input.as_ptr(), input.len());
+    pub fn update(&mut self, mut input: &[u8]) {
+        let mut block_state = AlignedBlockState::new(&mut self.opaque);
+        let block_state_buf = block_state.aligned_buf();
+
+        if self.buf_used != 0 {
+            let todo = core::cmp::min(input.len(), BLOCK_LEN - self.buf_used);
+
+            self.buf[self.buf_used .. self.buf_used + todo].copy_from_slice(&input[..todo]);
+            self.buf_used += todo;
+            input = &input[todo..];
+
+            if self.buf_used == BLOCK_LEN {
+                self.func.blocks(block_state_buf, &mut self.buf, 1 /* pad */);
+                self.buf_used = 0;
+            }
+        }
+
+        if input.len() >= BLOCK_LEN {
+            let todo = input.len() & !(BLOCK_LEN - 1);
+            let (complete_blocks, remainder) = input.split_at(todo);
+            self.func.blocks(block_state_buf, complete_blocks, 1 /* pad */);
+            input = remainder;
+        }
+
+        if input.len() != 0 {
+            self.buf[..input.len()].copy_from_slice(input);
+            self.buf_used = input.len();
         }
     }
 
-    #[inline]
     pub fn sign(mut self, tag_out: &mut Tag) {
-        unsafe { GFp_poly1305_finish(&mut self.state, tag_out); }
+        let mut block_state = AlignedBlockState::new(&mut self.opaque);
+        let block_state_buf = block_state.aligned_buf();
+
+        if self.buf_used != 0 {
+            self.buf[self.buf_used] = 1;
+            for byte in &mut self.buf[self.buf_used+1..] {
+                *byte = 0;
+            }
+            self.func.blocks(block_state_buf, &self.buf[..], 0 /* already padded */);
+        }
+
+        self.func.emit(block_state_buf, tag_out, &self.nonce);
```

Comment:
This is what I'm wondering about: Does any implementation of `emit` require the nonce to be aligned?